### DISCUSSION TITLE
angle between 2 3d vectors

### DISCUSSION THREAD

**Discussion by zeffii** [nortikin/sverchok#4495] [Ideas]
here's the partial solution to triangulating an arbitrary ngon, the main section i'll try to do tomorrow.. but this is already a useful function to have. And this video https://www.youtube.com/watch?v=QAdfkylpYwc
```python
"""
>in verts_in v
out verts_out v
out faces_out s
"""
import sverchok
from sverchok.utils.modules.vertex_utils import np_faces_normals as normals

def np_norm(A): return np.max(np.abs(A))
def ccw_angle(v1, v2, normal):
    """
    https://uk.mathworks.com/matlabcentral/answers/461410-how-to-get-direction-for-3d-angles-between-2-vectors
    """
    f = np.cross(v1, v2)
    ang = np.arctan2(np_norm(np.cross(v1, v2)), np.dot(v1, v2))
    if np.dot(np.cross(v1, v2), normal) < 0:
        ang = 2*np.pi - ang
    return np.degrees(ang)

def triangulate(verts):
    normal = normals(np.array([verts]))[0]
    print(normal)
    # num_faces = len(verts) - 2
    # while len(remaining) > 3:
    #     ...
    v_origin = np.array(verts[0])
    v1 = np.array(verts[1]) - v_origin
    v2 = np.array(verts[-1]) - v_origin 
    ang = ccw_angle(v1, v2, normal)
    print(ang)


for verts in verts_in:
    # print(verts)
    geom = triangulate(verts)
    # verts_out.append(geom[0])
    # faces_out.append(geom[1])
    
```

**Comment by Durman** [comment]
I guess build-in into Blender tessellation function should be faster https://docs.blender.org/api/3.1/mathutils.geometry.html#mathutils.geometry.tessellate_polygon

**Reply by zeffii** [reply]
yes, most probably.  but it can't be numba'd in `no_python` mode .. i think,

**Reply by Durman** [reply]
Ah, the weak point of the build-in function is that it can't handle multiple polygons efficiently. It would be nice to have a function which can take array of polygons and tessellate them. But with our polygon format I doubt that it's possible even with Numba because they can't be sored in NumPy arrays.

In Geometry Nodes they have another format for polygons which probably we can try to use. Polygons there consists of 3 arrays. First array points to the vertex indexes where the polygons start. Second array has number of vertices in a polygon. And last array is similar to our format but it's flat.
